internal/zinc-core/src/main/scala/sbt/internal/inc/InvalidationProfiler.scala

```diff
@@ -0,0 +1,267 @@
+package sbt.internal.inc
+
+import java.io.File
+
+import xsbti.UseScope
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+
+/**
+ * Defines a profiler interface that translates to the profiling protobuf schema.
+ *
+ * The protobuf schema has been mildly inspired from pprof:
+ * https://github.com/google/pprof/blob/master/proto/profile.proto
+ *
+ * A profiler interface should only be used by project, and not globally, as
+ * this class is not thread safe.
+ */
+abstract class InvalidationProfiler {
+  def profileRun: RunProfiler
+  def registerRun(run: zprof.ZincRun): Unit
+}
+
+object InvalidationProfiler {
+  final val empty: InvalidationProfiler = new InvalidationProfiler {
+    override def profileRun: RunProfiler = RunProfiler.empty
+    override def registerRun(run: zprof.ZincRun): Unit = ()
+  }
+}
+
+class ZincInvalidationProfiler extends InvalidationProfiler {
+  private final var lastKnownIndex: Long = -1L
+  private final val stringTable1: ArrayBuffer[String] = new ArrayBuffer[String](1000)
+  private final val stringTable2: ArrayBuffer[String] = new ArrayBuffer[String](10)
+  private final val stringTableIndices: mutable.HashMap[String, Long] =
```

Comment:
Can you add some information what these indices are for? `String` / `Long` are very generic types and people like me with very few compiler experience would find one sentence explaining the variable quite helpful :smile:

Reply:
What would you write? I didn't write anything because i thought the variable names were self-descriptive, i.e. `stringTableIndices` means a map from string to indices of the string table.

Reply:
what I don't understand (yet) is what these strings are. Class names, variable names, file paths or all of this :grin: And `long` is the index of what. 

If you say it's self explaining once one is deeper into how things work, I'm fine with not putting any docs on it :wink:

Reply:
No, I think you make a good point. It's likely that I'm too much into the implementation that I see it obvious, but it isn't. Thanks for pointing that out, I'll add a quick comment.